src/ast/type-updating.h

```diff
@@ -0,0 +1,282 @@
+  // unreachability
+  std::map<Expression*, Expression*> parents;
+
+  void visitExpression(Expression* curr) {
+    if (expressionStack.size() > 1) {
+      parents[curr] = expressionStack[expressionStack.size() - 2];
+    } else {
+      parents[curr] = nullptr; // this is the top level
+    }
+    // discover block/break relationships
+    if (auto* block = curr->dynCast<Block>()) {
+      if (block->name.is()) {
+        blockInfos[block->name].block = block;
+      }
+    } else if (auto* br = curr->dynCast<Break>()) {
+      // ensure info exists, discoverBreaks can then fill it
+      blockInfos[br->name];
+    } else if (auto* sw = curr->dynCast<Switch>()) {
+      // ensure info exists, discoverBreaks can then fill it
+      for (auto target : sw->targets) {
+        blockInfos[target];
+      }
+      blockInfos[sw->default_];
+    }
+    // add a break to the info, for break and switch
+    discoverBreaks(curr, +1);
+  }
+
+  // Part 2: Updating
+
+  // Node replacements, additions, removals and type changes should be noted. An
+  // exception is nodes you know will never be looked at again.
+
+  // note the replacement of one node with another. this should be called
+  // after performing the replacement.
+  // this does *not* look into the node. you should do so yourself if necessary
+  void noteReplacement(Expression* from, Expression* to) {
+    auto parent = parents[from];
+    noteRemoval(from);
+    // if we are replacing with a child, i.e. a node that was already present
+    // in the ast, then we just have a type and parent to update
+    if (parents.find(to) != parents.end()) {
+      parents[to] = parent;
+      if (from->type != to->type) {
+        propagateTypesUp(to);
+      }
+    } else {
+      noteAddition(to, parent, from);
+    }
+  }
+
+  // note the removal of a node
+  void noteRemoval(Expression* curr) {
+    noteRemovalOrAddition(curr, nullptr);
+    parents.erase(curr);
+  }
+
+  // note the removal of a node and all its children
+  void noteRecursiveRemoval(Expression* curr) {
+    struct Recurser : public PostWalker<Recurser, UnifiedExpressionVisitor<Recurser>> {
+      TypeUpdater& parent;
+
+      Recurser(TypeUpdater& parent, Expression* root) : parent(parent) {
+        walk(root);
+      }
+
+      void visitExpression(Expression* curr) {
+        parent.noteRemoval(curr);
+      }
+    };
+
+    Recurser(*this, curr);
+  }
+
+  void noteAddition(Expression* curr, Expression* parent, Expression* previous = nullptr) {
+    assert(parents.find(curr) == parents.end()); // must not already exist
+    noteRemovalOrAddition(curr, parent);
+    // if we didn't replace with the exact same type, propagate types up
+    if (!(previous && previous->type == curr->type)) {
+      propagateTypesUp(curr);
+    }
+  }
+
+  // if parent is nullptr, this is a removal
+  void noteRemovalOrAddition(Expression* curr, Expression* parent) {
+    parents[curr] = parent;
+    discoverBreaks(curr, parent ? +1 : -1);
```

Comment:
Mostly I'm not used to seeing unary no-op `+`, especially without the corresponding `-1`, particularly on this line https://github.com/WebAssembly/binaryen/blob/76af34609b2d3aeb18ec34275b3991663d9da742/src/ast/type-updating.h#L68.
My reading here is that you wrote it that way to tell the reader that the `1` is a delta, which is a good thing to communicate. The advantages I see of the enum idea is that
1. you can't lose the context across call frames. We pass the `int change` value around through a few functions, and without knowing how it's going to be used, it could have any int value.
2. you can more easily reason about it because we're constraining the range from "all int values" to "+1/-1", and encoding that information in the type.

The unary pluses are unusual, but if we don't want to go with an enum then I'd prefer we leave them in, because I think it does clue the reader in to the purpose of the value at the call site.